thunks.go

```diff
@@ -0,0 +1,169 @@
+		body: body,
+	})
+}
+
+func (t *thunk) getValue(i *interpreter, trace *TraceElement) (value, error) {
+	context := TraceContext{
+		Loc:  t.body.Loc(),
+		Name: "thunk <" + string(t.name) + ">",
+	}
+	return i.EvalInCleanEnv(trace, &context, &t.env, t.body)
+}
+
+// callThunk represents a concrete, but not yet evaluated call to a function
+type callThunk struct {
+	function evalCallable
+	args     callArguments
+}
+
+func makeCallThunk(ec evalCallable, args callArguments) *callThunk {
+	return &callThunk{function: ec, args: args}
+}
+
+func (th *callThunk) getValue(i *interpreter, trace *TraceElement) (value, error) {
+	evaluator := makeEvaluator(i, trace)
+	// TODO(sbarzowski): actually this trace is kinda useless inside...
+	return th.function.EvalCall(th.args, evaluator)
+}
+
+// cachedThunk is a wrapper that caches the value of a potentialValue after
+// the first evaluation.
+// Note: All potentialValues are required to provide the same value every time,
+// so it's only there for efficiency.
+// TODO(sbarzowski) better name?
+// TODO(sbarzowski) force use cached/ready everywhere? perhaps an interface tag?
+// TODO(sbarzowski) investigate efficiency of various representations
+type cachedThunk struct {
+	pv potentialValue
+}
+
+func makeCachedThunk(pv potentialValue) *cachedThunk {
+	return &cachedThunk{pv}
+}
+
+func (t *cachedThunk) getValue(i *interpreter, trace *TraceElement) (value, error) {
+	v, err := t.pv.getValue(i, trace)
+	if err != nil {
+		// TODO(sbarzowski) perhaps cache errors as well
+		// may be necessary if we allow handling them in any way
+		return nil, err
+	}
+	t.pv = &readyValue{v}
+	return v, nil
+}
+
+// errorThunk can be used when potentialValue is expected, but we already
+// know that something went wrong
+type errorThunk struct {
+	err error
+}
+
+func (th *errorThunk) getValue(i *interpreter, trace *TraceElement) (value, error) {
+	return nil, th.err
+}
+
+func makeErrorThunk(err error) *errorThunk {
+	return &errorThunk{err}
+}
+
+// unboundFields
+// -------------------------------------
+
+type codeUnboundField struct {
+	body astNode
+}
+
+func (f *codeUnboundField) bindToObject(sb selfBinding, origBinding bindingFrame) potentialValue {
+	// TODO(sbarzowski) nicer name
+	return makeThunk("object_field", makeEnvironment(origBinding, sb), f.body)
+}
+
+// evalCallables
+// -------------------------------------
+
+type closure struct {
+	// base environment of a closure
+	// arguments should be added to it, before executing it
+	env      environment
+	function *astFunction
+}
+
+func (closure *closure) EvalCall(arguments callArguments, e *evaluator) (value, error) {
+	argThunks := make(bindingFrame)
+	for i, arg := range arguments.positional {
+		argThunks[closure.function.parameters[i]] = arg
+	}
+
+	calledEnvironment := makeEnvironment(
+		addBindings(closure.env.upValues, argThunks),
+		closure.env.sb,
+	)
+	context := TraceContext{
+		Loc:  closure.function.Loc(),
+		Name: "function <anonymous>",
```

Comment:
Add TODO: better function names and object names

Reply:
done